### PR TITLE
Streamline trees' nodes hierarchy

### PR DESCRIPTION
**Pull request by smastelini**
This PR removes the repeated usage of multiple inheritances to handle active and inactive learning nodes.

We now see the concepts of "active" and "inactive" as states nodes might assume through time, instead of explicitly treating them as different node classes.

The benefits of it are clear most of the time: fewer lines of code to maintain, a simpler structure to understand. Future Cython implementations of the current tree module's members will also benefit from a simplified class structure.

Unfortunately, there is no free lunch. The ARF-based trees and EFDT modify the standard behavior of the learning nodes. As a result, I had to duplicate some code to ensure proper functionality. In all the cases, I preferred to repeat what these "fancy" decision tree algorithms modify in the vanilla Hoeffding Tree rather than reimplementing the learning logic of the nodes they inherit from. In the end, this choice might consume more lines of code, but I believe this is a safer option in the long run. If we fix/update something in the learning nodes of 
`HoeffdingTree{Classifier/Regressor}`, automatically `ARFHoeffdingTree{Classifier/Regressor}`  and EFDT will reflect those changes.

I cannot say I totally like that, to be fair. I would like not to repeat code at all. So, I am totally open to suggestions on how to solve this puzzle.

**Edit:** thanks to @jacobmontiel's suggestions, I was able to simplify the structures of the nodes from ARF and EFDT. I am happy with the results now :D

*Observation:* I changed the placement of the `learning_one` method in `HoeffdingTreeClassifier` just for aesthetic reasons. It made the git diff tracker crazy :P

### CODE REVIEW

**Comment by jacobmontiel** [comment]
During the refactoring, we removed the `sortedcontainers`. In the case of `best_split_suggestions ` now we get an unsorted list. We can use a heap (see [`heapq`](https://docs.python.org/3.8/library/heapq.html)) to avoid the overhead from sorting.

https://github.com/creme-ml/creme/blob/554e4c542fd085fb0fee1d3f0e0f0a57f71be71a/river/tree/_nodes/base.py#L384-L412

File: river/tree/_base_tree.py [hunk 1]
```diff
@@ -270,68 +268,30 @@ def _estimate_model_size(self):
         if actual_model_size > self._max_byte_size:
             self._enforce_size_limit()
 
-    def _deactivate_all_leaves(self):
+    def _activate_all_leaves(self):
         """Deactivate all leaves. """
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/tree/_base_tree.py:272]: Nice catch! In fact, on second thought it should be indeed removed. The trees already disable/enable nodes on demand.

File: river/tree/_nodes/arf_htc_nodes.py [hunk 2]
```diff
@@ -58,10 +34,37 @@ def __init__(self, initial_stats, depth, max_features, seed):
         self._rng = check_random_state(self.seed)
         self.feature_indices = []
 
+    def update_attribute_observers(self, x, y, sample_weight, tree, **kwargs):
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/tree/_nodes/arf_htc_nodes.py:37]: That would be my preferred choice. However, I am not sure how that will play out with the changes Wayne is performing.
I think I'll follow your suggested strategy and, in the future, we can check the consequences.

File: river/tree/_nodes/base.py [hunk 3]
```diff
@@ -40,30 +36,17 @@ class Node(metaclass=ABCMeta):
 
     Parameters
     ----------
-    stats
+    initial_stats
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/tree/_nodes/base.py:39]: You are right. I am going to standardize the usage of this parameter.

File: river/tree/hoeffding_tree_classifier.py [hunk 4]
```diff
@@ -143,6 +140,105 @@ def leaf_prediction(self, leaf_prediction):
...
+                split_criterion = HellingerDistanceCriterion()
+            else:
+                split_criterion = InfoGainSplitCriterion()
+            best_split_suggestions = node.best_split_suggestions(split_criterion, self)
+            best_split_suggestions.sort(key=attrgetter('merit'))
+            should_split = False
+            if len(best_split_suggestions) < 2:
+                should_split = len(best_split_suggestions) > 0
+            else:
+                hoeffding_bound = self._hoeffding_bound(split_criterion.range_of_merit(
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/tree/hoeffding_tree_classifier.py:196]: Should I put it in in `river.trees._tree_utils.py`?

**Comment by smastelini** [comment]
> During the refactoring, we removed the sortedcontainers. In the case of best_split_suggestions now we get an unsorted list. We can use a heap (see heapq) to avoid the overhead from sorting.

Just to be sure. Were lines such as the following previously redundant?

https://github.com/creme-ml/creme/blob/63120d370f01da0f44a45090b9b2ea55415ccbff/river/tree/hoeffding_tree_classifier.py#L287

I can use the bisect package to keep the splot suggestions ordered and alleviate the costs of sorting. What do you think?

File: river/tree/_nodes/base.py [hunk 5]
```diff
@@ -2,16 +2,13 @@
...
+from river.stats import Var
 from river.tree._attribute_test import InstanceConditionalTest
 from river.tree._attribute_test import AttributeSplitSuggestion
-from river.tree._attribute_observer import AttributeObserverNull
-
-
-NodeType = TypeVar('NodeType', bound='Node')
 
 
 class FoundNode:
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/tree/_nodes/base.py:14]: Nice catch! I will replace it by a namedtuple.

File: river/tree/_nodes/hatc_nodes.py [hunk 6]
```diff
@@ -86,7 +81,7 @@ def error_is_null(self):
     def kill_tree_children(self, hat):
         pass
 
-    def learn_one(self, x, y, sample_weight, tree, parent, parent_branch):
+    def learn_one(self, x, y, *, sample_weight=1., tree=None, parent=None, parent_branch=-1):
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/tree/_nodes/hatc_nodes.py:84]: That is a valid question. I did some small internal changes to reflect your question. For example, the attribute observer update only uses the list nominal attributes, so I changed the method's signature.

However, some trees rely on the actual Hoeffding tree, thus the need for the `tree` parameter. Take as examples the Hoeffding Adaptive Trees.

Hence, to make such a change, we will need to review our tree design once more. I think that this is something to check in the future.

File: river/tree/_nodes/efdtc_nodes.py [hunk 7]
```diff
@@ -87,15 +219,44 @@ class EFDTSplitNode(SplitNode, EFDTActiveLeaf):
...
 
+    @staticmethod
+    def new_nominal_attribute_observer(**kwargs):
+        return NominalAttributeClassObserver(**kwargs)
+
+    @staticmethod
+    def new_numeric_attribute_observer(**kwargs):
+        return NumericAttributeClassObserverGaussian(**kwargs)
+
+    @property
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/tree/_nodes/efdtc_nodes.py:233]: With the most recent refactoring, the split node inherited the property to access `_attribute_observer` from `LearningNode`, so this portion was removed.

File: river/tree/_base_tree.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (smastelini) [reply river/tree/_base_tree.py]: I removed this function in commit [63e5867](https://github.com/creme-ml/creme/pull/373/commits/63e58671f520a75c3292f709defe94a9cf6c19a8).

File: river/tree/_nodes/base.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (smastelini) [reply river/tree/_nodes/base.py]: It worked! Now FoundNode is a namedtuple (commit [63e5867](https://github.com/creme-ml/creme/pull/373/commits/63e58671f520a75c3292f709defe94a9cf6c19a8)).

File: river/tree/_nodes/arf_htc_nodes.py [hunk 8]
```diff
@@ -102,9 +131,34 @@ class RandomActiveLearningNodeNBA(LearningNodeNBA, RandomActiveLeafClass):
...
+            try:
+                obs = self.attribute_observers[idx]
+            except KeyError:
+                if ((tree.nominal_attributes is not None and idx in tree.nominal_attributes)
+                        or not isinstance(x[idx], numbers.Number)):
+                    obs = self.new_nominal_attribute_observer(**kwargs)
+                else:
+                    obs = self.new_numeric_attribute_observer(**kwargs)
+                self.attribute_observers[idx] = obs
+            obs.update(x[idx], y, sample_weight)
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/tree/_nodes/arf_htc_nodes.py]: Added the missing empty line in commit [63e5867](https://github.com/creme-ml/creme/pull/373/commits/63e58671f520a75c3292f709defe94a9cf6c19a8).

File: river/tree/_nodes/arf_htc_nodes.py [hunk 9]
```diff
@@ -58,10 +34,37 @@ def __init__(self, initial_stats, depth, max_features, seed):
         self._rng = check_random_state(self.seed)
         self.feature_indices = []
 
+    def update_attribute_observers(self, x, y, sample_weight, tree, **kwargs):
+        def sample_features(instance, rng, max_features):
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/tree/_nodes/arf_htc_nodes.py]: By refactoring the nodes once more, I changed the internal structure. It is looking better now :P

The changes were implemented in commit [63e5867](https://github.com/creme-ml/creme/pull/373/commits/63e58671f520a75c3292f709defe94a9cf6c19a8).

File: river/tree/_nodes/arf_htc_nodes.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (smastelini) [reply river/tree/_nodes/arf_htc_nodes.py]: I finally refactored the missing parts and simplified them. You check the changes in commit [63e5867](https://github.com/creme-ml/creme/pull/373/commits/63e58671f520a75c3292f709defe94a9cf6c19a8).

File: river/tree/_nodes/arf_htr_nodes.py [hunk 10]
```diff
@@ -1,59 +1,18 @@
...
-        for obs in self.attribute_observers.values():
-            if isinstance(obs, NumericAttributeRegressionObserver):
-                obs.remove_bad_splits(criterion=criterion, last_check_ratio=last_check_ratio,
-                                      last_check_vr=last_check_vr, last_check_e=last_check_e,
-                                      pre_split_dist=self.stats)
-
-
-class RandomActiveLearningNodeMean(LearningNodeMean, RandomActiveLeafRegressor):
-    """ Learning Node for regression tasks that always use the average target
+class RandomLearningNodeMean(LearningNodeMean):
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/tree/_nodes/arf_htr_nodes.py]: These nodes were refactored once more in commit [63e5867](https://github.com/creme-ml/creme/pull/373/commits/63e58671f520a75c3292f709defe94a9cf6c19a8).

File: river/tree/_nodes/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (smastelini) [reply river/tree/_nodes/base.py]: You are right. In commit [63e5867](https://github.com/creme-ml/creme/pull/373/commits/63e58671f520a75c3292f709defe94a9cf6c19a8) the parameter name was standardized in the nodes.

File: river/tree/_nodes/efdtc_nodes.py [hunk 11]
```diff
@@ -71,8 +73,138 @@ def count_nodes():
         return Counter(leaf_nodes=1, decision_nodes=0)
 
 
-class EFDTSplitNode(SplitNode, EFDTActiveLeaf):
-    """Node that splits the data in a EFDT.
+class EFDTLearningNodeNB(LearningNodeNB):
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/tree/_nodes/efdtc_nodes.py]: This one took me some time, due to the exotic nature of the nodes' learning process. But now they are refactored in commit [63e5867](https://github.com/creme-ml/creme/pull/373/commits/63e58671f520a75c3292f709defe94a9cf6c19a8).

File: river/tree/hoeffding_adaptive_tree_regressor.py [hunk 12]
```diff
@@ -214,28 +204,15 @@ def _new_learning_node(self, initial_stats=None, parent=None, is_active=True):
...
-                initial_stats=initial_stats, depth=depth, leaf_model=leaf_model,
-                adwin_delta=self.adwin_confidence, seed=self.seed)
+        new_ada_leaf = AdaLearningNodeRegressor(
+            initial_stats=initial_stats, depth=depth, leaf_model=leaf_model,
+            adwin_delta=self.adwin_confidence, seed=self.seed)
 
-            if parent is not None and not isinstance(parent, SplitNode):
-                new_ada_leaf._fmse_mean = parent._fmse_mean
-                new_ada_leaf._fmse_model = parent._fmse_model
+        if parent is not None and not isinstance(parent, SplitNode):
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/tree/hoeffding_adaptive_tree_regressor.py]: Could you please elaborate?

**Comment by smastelini** [comment]
Each insertion in a list already ordered takes `O(log n)`, thus `O(n log n)` for inserting `n` elements. Python's TimSort also has a cost of `O(n log n)`.

Therefore, I think we are in good shape since the trees call `sort` in the split suggestions.

File: river/tree/_nodes/hatc_nodes.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jacobmontiel) [reply river/tree/_nodes/hatc_nodes.py:84]: Thanks for looking into this. I agree, this is out of the scope of the PR.

File: river/tree/hoeffding_adaptive_tree_regressor.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (jacobmontiel) [reply river/tree/hoeffding_adaptive_tree_regressor.py]: This part of the code was inside an `if` block: 

`if is_active:`

**Comment by jacobmontiel** [comment]
> Each insertion in a list already ordered takes `O(log n)`, thus `O(n log n)` for inserting `n` elements. Python's TimSort also has a cost of `O(n log n)`.
> 
> Therefore, I think we are in good shape since the trees call `sort` in the split suggestions.

Thanks for the follow up on this.

File: river/tree/hoeffding_adaptive_tree_regressor.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (smastelini) [reply river/tree/hoeffding_adaptive_tree_regressor.py]: Oh, I see. The if-block was valid when we had separate Active and Inactive node implementations. Now, that is not the case anymore :)

[the remark this answers is not in the document]
Comment (jacobmontiel) [reply river/tree/hoeffding_adaptive_tree_regressor.py]: OK, thanks for the clarification.

File: river/tree/hoeffding_tree_classifier.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jacobmontiel) [reply river/tree/hoeffding_tree_classifier.py]: It has been agreed to leave the current implementation.

**Review by jacobmontiel — APPROVED** [review]
Awesome work @smastelini ! This is a nice improvement in the sub-package.

The PR is ready to merge.

**Pull request merged**